File: php/daily_report/daily_report.php
<?php 

function reportFromFile($fileName) {
	$report = array();
	$total_bytes=0;
	$total_requests=0;

	$handle = @fopen($fileName, "r");
	if ($handle) {
	    while (($line = fgets($handle)) !== false) {
	        list($timestamp,$bytes,$status,$ip_addr) = explode(";", $line);
	        if($status != "200 OK") 
	        	continue;

	        $ip_addr = trim($ip_addr);
			
	        if($report[$ip_addr]) {
	        	$report[$ip_addr]["requests"]++;
	        	$report[$ip_addr]["bytes"]+=(int)$bytes;
	        } else {
				$report[$ip_addr] = array("requests" => 1, "bytes" => (int)$bytes);
	        }

	        $total_requests++;
			$total_bytes += (int)$bytes;
	    }

	    fclose($handle);
	
		$callback = function($a, $b) {
		    if ($a["requests"] == $b["requests"]) {
		        return 0;
		    }
		    return ($a["requests"] > $b["requests"]) ? -1 : 1;
		};
		
		uasort($report, $callback);
		echo "IP Address;Requests;Total Requests %;Bytes;Total Bytes %\n";
		foreach ($report as $ip => $data) {
			echo $ip.";"
				.$data["requests"].";".($data["requests"] * 100.0 / $total_requests).";"
				.$data["bytes"].";".($data["bytes"] * 100.0 / $total_bytes)."\n";
		}
	} else {
	    echo "Error opening file!";
	} 
}
?>

File: java/src/main/java/net/cristcost/data/dailyreport/DailyReport.java
package net.cristcost.data.dailyreport;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintStream;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

public class DailyReport {

  public class RequestEntry {

    private int requests = 0;
    private int bytes = 0;

    public RequestEntry(int bytes, int requests) {
      this.requests = requests;
      this.bytes = bytes;
    }

    public void logRequest(int newBytes) {
      requests++;
      bytes += newBytes;
    }
  }

  private Map<String, RequestEntry> entries = new HashMap<>();

  public int totalRequests = 0;
  public int totalBytes = 0;

  public void parse(String fileName) throws IOException {
    try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
      String line;
      while ((line = br.readLine()) != null) {
        parseLine(line);
      }
    }
  }

  public void print(PrintStream out) {
    printHeader(out);
    for (Entry<String, RequestEntry> entry : entries.entrySet()) {

      int requests = entry.getValue().requests;
      int bytes = entry.getValue().bytes;

      StringBuilder sb = new StringBuilder();
      sb.append(entry.getKey());
      sb.append(",");
      sb.append(requests);
      sb.append(",");
      sb.append(requests * 100.0 / totalRequests);
      sb.append(",");
      sb.append(bytes);
      sb.append(",");
      sb.append(bytes * 100.0 / totalBytes);

      out.println(sb.toString());
    }
  }

  private void printHeader(PrintStream out) {
    System.out.println("IP Address,Number of Requests,Percentage,Total Bytes, Percentage");
  }

  private void logRequest(String remoteAddress, int bytes) {
    if (entries.containsKey(remoteAddress)) {
      entries.get(remoteAddress).logRequest(bytes);
    } else {
      entries.put(remoteAddress, new RequestEntry(bytes, 1));
    }
  }

  private void parseLine(String line) {
    String[] split = line.split(";");
    parseLine(split[0], split[1], split[2], split[3]);
  }

  private void parseLine(String timestamp, String bytesAsString, String status,
      String remoteAddress) {
    if (status.equals("200 OK")) {
      int bytes = Integer.parseInt(bytesAsString);
      totalBytes += bytes;
      totalRequests++;
      logRequest(remoteAddress, bytes);
    }
  }
}